packages/base/README.md

```diff
@@ -1,2 +1,94 @@
-# appium-base-plugin
-The base class used to create Appium 2.0 plugins
+# Appium Base Plugin
+
+The base class used to create Appium plugins. This plugin should not be installed directly as it does nothing. Instead, you extend this plugin when creating your *own* Appium plugins.
+
+## Steps to create your own plugin
+
+You can follow this guide to create your own Appium plugin.
+
+### 1. Initialize a Node.js project
+
+Appium plugins are Node.js programs, though of course you can call out to any other language or framework you want in the course of executing Node.js code. The basic requirement for an Appium plugin is to add certain metadata fields in your project's `package.json`, for example:
+
+```
+{
+  ...,
+  "appium": {
+    "pluginName": "your-plugin-name",
+    "mainClass": "YourPlugin",
+  },
+  ...
+}
+```
+
+Here the `pluginName` is the name used to activate your plugin by users who have installed it. So if your `pluginName` is `cool-plugin`, they would activate it as follows:
+
+```
+appium --plugins=cool-plugin
+```
+
+The same goes for updating and removing the plugin using the CLI.
+
+The `mainClass` value is the name that Appium should import from your project's entrypoint. So for example if somewhere in your project you have defined your plugin as follows:
+
+```
+class CoolPlugin extends BasePlugin {
+  ...
+}
+```
+
+Then you need to make sure that Appium can import/require this name from your node package:
+
+```
+const { CoolPlugin } = require('someones-cool-plugin')
+```
+
+### 2. Create your Plugin class by extending BasePlugin
+
+As implied above, your project should export a class which extends BasePlugin. This means you need to add BasePlugin as a project dependency:
+
+```
+npm install --save @appium/base-plugin
+```
+
+Then you can create a class which extends it:
+
+```
+import BasePlugin from '@appium/base-plugin'
+
+class CoolPlugin extends BasePlugin {}
+```
+
+### 3. Implement your Plugin class
+
+Your plugin can do one or more of 3 basic things:
+
+1. Update the Appium server object before it starts listening for requests
+2. Handle Appium commands in lieu of the driver which would normally handle them
+3. Add new routes/commands to Appium (which your plugin must then handle)
+
+#### Updating the server
+
+You probably don't normally need to update the Appium server object (which is an Express server having already been configured in a variety of ways). But, for example, you could add new Express middleware to the server to support your plugin's requirements. To update the server you must do two things:
+
+1. Set the class variable `updatesServer` to `true`, so that Appium knows your plugin needs to update the server before launch.
+2. Implement the `async updateServer` method in your class. This method takes two parameters, `expressApp` and `httpServer`, which are the Express app object and the Node HTTP server object respectively. You can do whatever you want with them inside the `updateServer` method.
```

Comment:
> Wouldn't it be enough to check for an updateServer in the plugin prototype and if so assume the user likes to update the server?

Not currently because the base plugin has `updateServer` defined, though granted this is primarily for the purpose of hosting docstrings.

Reply:
> Having the flexibility to work with the express objects is great no question but if Appium can simplify the most common use case it could make things easier!

Implementing new commands doesn't require working with express objects, it involves handling it as part of `handle`. But I think you're right we could come up with a convention where we say a plugin handles a command if a method with that command name exists on the plugin instance. I think I like that. I'll probably implement it as a separate set of changes between this repo and basedriver.

Reply:
> A question to be answered here is what happens if a command/endpoint is already defined? I assume right now it would just execute both command implementations but I can see scenarios where this is not a desired behavior or can cause confusion.

This is why we have `next` provided as part of `handle`. It allows a plugin to wrap the default behavior, or execute it after its own functionality. Here `next` encodes a whole chain of plugins that might act on a command. But it's up to each plugin to play nice, because we can't assume that plugins are designed to work with one another. If the goal of  plugin A is to completely replace the functionality of a command, and the goal of plugin B is the same, then whichever plugin is loaded first is the one which will handle it, and the other won't.

This does mean that plugins are not necessarily compatible with one another and users might need to be aware of this. But I can't think of any meaningful way for us to force plugins to play nice with one another since the point is for them to have a lot of power.